system-addon/content-src/components/TopSites/TopSite.jsx

```diff
@@ -8,59 +8,125 @@ import {
-  } else {
-    // styles and class names for top sites with screenshot + small icon in top left corner
-    imageClassName = `screenshot${link.screenshot ? " active" : ""}`;
-    imageStyle = {backgroundImage: link.screenshot ? `url(${link.screenshot})` : "none"};
+export class TopSiteLink extends React.PureComponent {
+  constructor(props) {
+    super(props);
+    this.onDragStart = this.onDragStart.bind(this);
+    this.onDragEnd = this.onDragEnd.bind(this);
+    this.onDragOver = this.onDragOver.bind(this);
+    this.onDragEnter = this.onDragEnter.bind(this);
+    this.onDragLeave = this.onDragLeave.bind(this);
+    this.onDrop = this.onDrop.bind(this);
+  }
 
-    // only show a favicon in top left if it's greater than 16x16
-    if (faviconSize >= MIN_CORNER_FAVICON_SIZE) {
-      showSmallFavicon = true;
-      smallFaviconStyle = {backgroundImage:  `url(${link.favicon})`};
-    } else if (link.screenshot) {
-      // Don't show a small favicon if there is no screenshot, because that
-      // would result in two fallback icons
-      showSmallFavicon = true;
-      smallFaviconFallback = true;
+  /*
+   * Helper to determine whether the drop zone should allow a drop. We only allow
+   * dropping top sites for now.
+   */
+  _allowDrop(e, index) {
+    let draggedIndex = parseInt(e.dataTransfer.getData("text/topsite-index"), 10);
+    if (!isNaN(draggedIndex) && draggedIndex !== index) {
+      return true;
     }
+    return false;
   }
-  return (<li className={topSiteOuterClassName} key={link.guid || link.url}>
-   <a href={link.url} onClick={props.onClick}>
-      <div className="tile" aria-hidden={true} data-fallback={letterFallback}>
-        <div className={imageClassName} style={imageStyle} />
-        {showSmallFavicon && <div
-          className="top-site-icon default-icon"
-          data-fallback={smallFaviconFallback && letterFallback}
-          style={smallFaviconStyle} />}
-     </div>
-     <div className={`title ${link.isPinned ? "pinned" : ""}`}>
-       {link.isPinned && <div className="icon icon-pin-small" />}
-        <span dir="auto">{title}</span>
-     </div>
-   </a>
-   {props.children}
-  </li>);
-};
+  onDragStart(e) {
+    e.dataTransfer.effectAllowed = "move";
+    e.dataTransfer.setData("text/topsite-index", this.props.index);
+    e.target.blur();
+    this.props.onDragStart(this.props.index, this.props.link, this.props.title);
+  }
+  onDragEnd(e) {
+    this.props.onDragEnd();
+  }
+  onDragOver(e) {
+    if (this._allowDrop(e, this.props.index)) {
+      e.preventDefault();
+    }
+  }
+  onDragEnter(e) {
+    if (this._allowDrop(e, this.props.index)) {
+      e.preventDefault();
+      this.props.onDragEnter(this.props.index);
+    }
+  }
+  onDragLeave(e) {
+    if (this._allowDrop(e, this.props.index)) {
+      this.props.onDragLeave(this.props.index);
+    }
+  }
+  onDrop(e) {
+    if (this._allowDrop(e, this.props.index)) {
+      e.preventDefault();
+      this.props.onDrop(this.props.index);
```

Comment:
Although with a common `onEvent` could do…
```js
onEvent(event) {
  switch (event.type) {
    case "dragover":
    case "dragenter":
    case "dragleave":
    case "drop":
      if (this._allowDrop(event, index)) {
        event.preventDefault();
        this.props.onEvent(event, index);
```

? maybe…

Reply:
I think that's better? 422ea358690cab7217b3fc586ccabb5a85032ab6

Reply:
(except I broke something)